src/login.ts

```diff
@@ -6,52 +6,99 @@ import { MastoConfig } from './config';
-  /** Log level to print to your console */
-  readonly logLevel?: LogType;
-  /** Access token of your account */
-  readonly accessToken?: string;
   /** Timeout in milliseconds */
   readonly timeout?: number;
   readonly defaultRequestInit?: Omit<RequestInit, 'body' | 'method'>;
-  readonly disableVersionCheck?: boolean;
   readonly disableDeprecatedWarning?: boolean;
-};
+}
 
-export const login = async (params: LoginParams): Promise<Client> => {
-  const draft: Writable<MastoConfigProps> = {
+const buildConfigProps = (params: RequestParams): MastoConfigProps => {
+  return {
     url: params.url,
     streamingApiUrl: '',
-    logLevel: params.logLevel,
-    accessToken: params.accessToken,
     timeout: params.timeout,
     defaultRequestInit: params.defaultRequestInit,
-    disableVersionCheck: params.disableVersionCheck,
     disableDeprecatedWarning: params.disableDeprecatedWarning,
   };
+};
+
+const buildRequestContext = (
+  props: MastoConfigProps,
+): {
+  serializer: SerializerNativeImpl;
+  config: MastoConfig;
+  http: HttpNativeImpl;
+} => {
   const serializer = new SerializerNativeImpl();
+  const config = new MastoConfig(props, serializer);
+  const http = new HttpNativeImpl(serializer, config);
+  return { serializer, config, http };
+};
 
-  {
-    const config = new MastoConfig(draft, serializer);
-    const http = new HttpNativeImpl(serializer, config);
-    const instance = await new InstanceRepository(http, config).fetch();
-    draft.version = new SemVer(instance.version);
-    draft.streamingApiUrl = instance.urls.streamingApi;
-  }
+export const fetchV1Instance = (
+  params: RequestParams,
+): Promise<v1.Instance> => {
+  const { http, config } = buildRequestContext(buildConfigProps(params));
+  return new V1InstanceRepository(http, config).fetch();
+};
 
-  const config = new MastoConfig(draft, serializer);
+export const fetchV2Instance = (
+  params: RequestParams,
+): Promise<v2.Instance> => {
+  const { http, config } = buildRequestContext(buildConfigProps(params));
+  return new V2InstanceRepository(http, config).fetch();
+};
+
+export interface LoginParams extends RequestParams {
+  readonly streamingApiUrl: string;
+  readonly version?: string;
+  /** Log level to print to your console */
+  readonly logLevel?: LogType;
+  /** Access token of your account */
+  readonly accessToken?: string;
+  readonly disableVersionCheck?: boolean;
+}
+
+export const createClient = (params: LoginParams): Client => {
+  const props: Writable<MastoConfigProps> = {
+    ...buildConfigProps(params),
+    streamingApiUrl: params.streamingApiUrl,
+    logLevel: params.logLevel,
+    accessToken: params.accessToken,
+    disableVersionCheck: params.disableVersionCheck,
+    version: params.version ? new SemVer(params.version) : undefined,
+  };
+  const { serializer, http, config } = buildRequestContext(props);
   const logger = new LoggerConsoleImpl(config.getLogLevel());
   const ws = new WsNativeImpl(config, serializer);
```

Comment:
`ws` and `logger` is used only in `createClient` function. Do you have any idea about a better name?

Reply:
Ah, so you're not including it to reduce unneeded dependency to `ws` or `logger` from `fetch{v1,v2}Instance`? It makes sense in that case.

I conjure up a DI container by the name "context", so I thought it was natural to include all dependencies in the `RequestContext` interface. How about  `HttpContext` or something because the main concern here is to make an HTTP request.